feat(api): implement the liquid transfer function

sanni-t opened the pull request:
Closes AUTH-866

# Overview

Final PR of the 3-PR series.
Implements `InstrumentCore.transfer_liquid()`.

The `transfer_liquid()` method does the following:
1. Loads the relevant liquid class record into protocol engine, gets the liquid class ID back. This will be used in command annotations in the future
2. Breaks down the total transfer volume into piecewise transfers, if necessary.
3. Does auto-tip handling, unless the user has specified to 'never' use a new tip.
4. Calls `aspirate_liquid_class()` and `dispense_liquid_class()` for every source-> destination transfer.

This PR also updates the tiprack 'names' to URIs in the liquid class definition in shared data for 'water' only. We will have a separate PR for updating rest of the definitions.

## Test Plan and Hands on Testing

- Added some unit tests for things that are unit-testable
- Added integration tests for verifying that `transfer_liquid()` execution succeeds for various supported configurations
- TODO: on-robot testing

Additionally, we will be adding analysis snapshot tests with protocols that use various configurations for transferring with liquid classes.

## Review requests

- The implementation of `transfer_liquid()` turned out to be not easily unit-testable, mainly owing to the location of implementation (inside `InstrumentCore`). I plan on moving this function out of the InstrumentCore so that it can be unit-tested without the constraints of rest of the Instrument core. But this still serves as a robust implementation for doing end-to-end, integration testing so that's what I am focusing on in this PR.
- This PR also makes liquid-classes-based transfer available for hardware testing and verification.

## Risk assessment

Low. Doesn't change existing API

sanni-t committed:
initial changes

sanni-t committed:
added transfer liquid implementation

sanni-t committed:
added a pipette api-load-name getter

sanni-t committed:
code and type fixes

sanni-t committed:
cleanup & fixes

sanni-t committed:
fixed small bugs

sanni-t committed:
updated tiprack names to URI names

sanni-t committed:
don't air gap if air gap volume is 0

sanni-t committed:
added an integration test

ddcc4 commented on api/src/opentrons/protocol_api/core/engine/instrument.py:
```diff
@@ -705,6 +712,20 @@ def get_pipette_name(self) -> str:
+            pip_api_name
+            for pip_api_name, pip_name in PIPETTE_API_NAMES_MAP.items()
+            if pip_name == pipette.pipetteName
+        ][0]
```
Small nitpick: This can be more efficiently implemented as:
```
load_name = next(
  (pip_api_name
   for pip_api_name, pip_name in PIPETTE_API_NAMES_MAP.items()
   if pip_name == pipette.pipetteName),
  None
)
assert load_name
return load_name
```

ddcc4 commented on api/src/opentrons/protocol_api/core/engine/instrument.py:
```diff
@@ -883,7 +904,7 @@ def load_liquid_class(
         liquid_class_record = LiquidClassRecord(
             liquidClassName=liquid_class.name,
-            pipetteModel=self.get_model(),  # TODO: verify this is the correct 'model' to use
+            pipetteModel=pipette_load_name,  # TODO: verify this is the correct 'model' to use
```
Does comment still apply?

sanni-t replied:
Nope. I'll remove it.

ddcc4 commented on api/src/opentrons/protocol_api/core/engine/instrument.py:
```diff
@@ -928,54 +960,135 @@ def transfer_liquid(
+            pipette_load_name=self.get_pipette_load_name(),
+            tiprack_uri=tiprack_uri_for_transfer_props,
         )
         transfer_props = liquid_class.get_for(
```
`self.load_liquid_class()` above also calls `liquid_class.get_for()`, right?

Can you refactor `self.load_liquid_class()` to just take in the `transfer_props` directly as an argument?

ddcc4 commented on api/src/opentrons/protocol_api/core/engine/instrument.py:
```diff
@@ -1034,7 +1153,9 @@ def aspirate_liquid_class(
         components_executor.retract_after_aspiration(volume=volume)
-        return components_executor.tip_state.last_liquid_and_air_gap_in_tip
+        last_contents = components_executor.tip_state.last_liquid_and_air_gap_in_tip
+        tip_contents[-1] = last_contents
```
Hm, I'm a little confused what this is trying to do: it sometimes modifies the list that's passed in, and it sometimes modifies a list that was created local to this function?

sanni-t replied:
Talked about it in person. The list that's passed in is not re-used by the caller function so it doesn't matter if the list gets modified here or not, since the list in the caller function gets replaced by ones returned by the `aspirate_liquid_class` and `dispense_liquid_class`. But I can see how it can cause confusion so I'll update this to use a copy of the list for any modifications.

sanni-t committed:
address comments on PR-17138

sanni-t committed:
addressed comments

sanni-t committed:
added tests

sanni-t committed:
include liquid classes directory in shared data bundle

sanni-t committed:
fixed blowout location for source

ddcc4 committed:
Clean up transfer_liquid implementation for argument-passing hygiene (#17244)

- Functions should not modify any argument that's passed in (unless that's the purpose of the function).
- Objects should not modify any arguments given to their constructor (unless the object is supposed to take ownership of the argument).

sanni-t committed:
fixes for blowout and touch tip locations, dont add air gap in last step if not using new tip

sanni-t commented on api/tests/opentrons/protocol_api/core/engine/test_transfer_components_executor.py:
```diff
@@ -613,28 +615,31 @@ def test_retract_after_dispense_with_blowout_in_source(
+        mock_instrument_core.prepare_to_aspirate(),
         mock_instrument_core.air_gap_in_place(
             volume=air_gap_volume, flow_rate=air_gap_volume
-        ),
-        mock_instrument_core.delay(0.2),
+        )
+        if add_final_air_gap is True
+        else None,
+        mock_instrument_core.delay(0.2) if add_final_air_gap is True else None,
```
Not sure if this is the right way to test this. Right now the test is failing in an unexpected way.

ddcc4 committed:
Silly Python syntax (#17256)

Argument unpacking to make `decoy.verify()` work with optional
calls.

sanni-t committed:
updated other tests with blowout

sanni-t committed:
formatting

sanni-t committed:
fix validation test

jbleon95 reviewed (approved):
It all looks good to me, and tested this on the robot so we're ready to hand this off for hardware testing

ddcc4 commented on api/src/opentrons/protocol_api/core/engine/instrument.py:
```diff
@@ -928,54 +969,138 @@ def transfer_liquid(
+        ]
+        for step_volume, src_dest_combo in source_dest_per_volume_step:
+            step_source, step_destination = src_dest_combo
+            is_last_step = len([source_dest_per_volume_step]) > 0
```
How does this work? Why does `len([source_dest_per_volume_step]) > 0` mean it's the last step? And wouldn't the length **always** be > 0?

sanni-t replied:
Oh shoot! I meant to fix this! Thanks for catching it!

ddcc4 commented on api/src/opentrons/protocol_api/core/engine/transfer_components_executor.py:
```diff
@@ -50,36 +52,35 @@ class TipState:
         assert (
             self.last_liquid_and_air_gap_in_tip.air_gap == 0
         ), "Air gap present in the tip."
         self.last_liquid_and_air_gap_in_tip.liquid -= volume
```
Is it allowed to make `last_liquid_and_air_gap_in_tip.liquid` go below zero here?

sanni-t replied:
Nope. I can add an assertion there to make sure we never remove more liquid than is present

ddcc4 commented on shared-data/python/opentrons_shared_data/pipette/types.py:
```diff
@@ -252,6 +251,31 @@ class PipetteNameType(str, enum.Enum):
+# Mapping of public Python Protocol API pipette load names
+# to names used by the internal Opentrons system
+# TODO (spp, 2025-01-02): make the API load name a part of the pipette's definition
+PIPETTE_API_NAMES_MAP = {
```
Is this same dict in `api/src/opentrons/protocol_api/validation.py` as well? (I also needed it for Python generation from PD -- see [here](https://github.com/Opentrons/opentrons/compare/edge...dc-pd-pyexport#diff-56b2610043ec9c924442d74d00d42cbf471bd292a1d4ea79a009ec7914623d70R153).)

sanni-t replied:
Yep. I've moved it to shared data for precisely this reason that PD will need to use it too

sanni-t committed:
fix the last step checker logic

ddcc4 commented on api/src/opentrons/protocol_api/core/engine/instrument.py:
```diff
@@ -1043,9 +1043,19 @@ def _pick_up_tip() -> None:
+                    source_dest_per_volume_step
+                )
+            except StopIteration:
+                is_last_step = True
```
Is `source_dest_per_volume_step` an iterator or a list?
Can you turn it into a list?
If so, could you just `enumerate` over the list and check if you're at the index of the last item?

Like,
```
for idx, (next_step_volume, next_src_dest_combo) in enumerate(source_dest_per_volume_step):
  is_last_step = (idx == len(source_dest_per_volume_step) - 1)
  do stuff ..
  
```

sanni-t replied:
It's an iterator.
Yes I can turn it into a list but that defeats the purpose of having `expand_for_volume_constraints` as a generator function that uses memory more efficiently. The iterator can have quite a number of elements on some occasions. So unless there's a compelling reason to not do it the way I have over here, I think we should avoid converting it to list.

ddcc4 replied:
Ah I see. In that case, I wouldn't do this inline in the code here -- it's too much logic to pack in here. Instead, I would define a helper function that implements **another** generator to wrap `expand_for_volume_constraints()` that includes whether we're at the last step.

But you also control `expand_for_volume_constraints()`, don't you? Could you just make it include `is_last_step` in the values that it yields?

sanni-t committed:
resolved linter errors
